module/modmanager/manager.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	return nil
+}
+
+// IsModularService returns true if a service would be handled by a module.
+func (mgr *Manager) IsModularService(cfg config.Service) bool {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	return mgr.getServiceModule(cfg) != nil
+}
+
+// ReconfigureService updates/reconfigures a modular service with a new configuration.
+func (mgr *Manager) ReconfigureService(ctx context.Context, cfg config.Service) error {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	module := mgr.getServiceModule(cfg)
+	if module == nil {
+		return errors.Errorf("no module registered to serve resource api %s and model %s", cfg.ResourceName().Subtype, cfg.Model)
+	}
+
+	cfgProto, err := config.ServiceConfigToProto(&cfg)
+	if err != nil {
+		return err
+	}
+	_, err = module.client.ReconfigureService(ctx, &pb.ReconfigureServiceRequest{Config: cfgProto})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsModularResource returns true if a component is handled by a module.
+func (mgr *Manager) IsModularResource(name resource.Name) bool {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	_, ok := mgr.rMap[name]
+	return ok
+}
+
+// RemoveResource requests the removal of a resource from a module.
+func (mgr *Manager) RemoveResource(ctx context.Context, name resource.Name) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	module, ok := mgr.rMap[name]
+	if !ok {
+		return errors.Errorf("resource %+v not found in module", name)
+	}
+
+	_, err := module.client.RemoveResource(ctx, &pb.RemoveResourceRequest{Name: name.String()})
+	return err
+}
+
+func (mgr *Manager) getComponentModule(cfg config.Component) *module {
+	for _, module := range mgr.modules {
+		var api resource.RPCSubtype
+		var ok bool
+		for a := range module.handles {
+			if a.Subtype == cfg.ResourceName().Subtype {
+				api = a
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			continue
+		}
+		for _, model := range module.handles[api] {
+			if cfg.Model == model {
+				return module
+			}
+		}
+	}
+	return nil
+}
+
+func (mgr *Manager) getServiceModule(cfg config.Service) *module {
+	for _, module := range mgr.modules {
+		var api resource.RPCSubtype
+		var ok bool
+		for a := range module.handles {
+			if a.Subtype == cfg.ResourceName().Subtype {
+				api = a
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			continue
+		}
+		for _, model := range module.handles[api] {
+			if cfg.Model == model {
+				return module
+			}
+		}
+	}
+	return nil
+}
+
+// DepsToNames converts a dependency list to a simple string slice.
+func DepsToNames(deps registry.Dependencies) []string {
+	var depStrings []string
+	for dep := range deps {
+		depStrings = append(depStrings, dep.String())
+	}
+	return depStrings
+}
+
+func (m *module) checkReady(ctx context.Context, addr string) error {
+	ctxTimeout, cancelFunc := context.WithTimeout(ctx, time.Second*10)
+	defer cancelFunc()
+	for {
+		req := &pb.ReadyRequest{ParentAddress: addr}
+		resp, err := m.client.Ready(ctxTimeout, req, grpc_retry.WithMax(5000))
```

Comment:
5000 ms? if yes, can you make it a const somewhere

Reply:
No, it's "5000 attempts"... just an arbitrarily high number not meant to be hit (and this specific number was just kept from an example.) The real timeout is handled by the context timeout, but if WithMax isn't set (or is set to 0) the retry logic doesn't engage.

Added a comment to note the reason for the number.